onmt/modules/CudnnEncoder.lua

```diff
@@ -0,0 +1,134 @@
+--[[ CudnnEncoder is a unidirectional/bidirectional Encoder used for the source language
+  using CuDNN RNN implementation
+
+    h_1 => h_2 => h_3 => ... => h_n
+     |      |      |             |
+     .      .      .             .
+     |      |      |             |
+    h_1 => h_2 => h_3 => ... => h_n
+     |      |      |             |
+     |      |      |             |
+    x_1    x_2    x_3           x_n
+
+--]]
+local CudnnEncoder, parent = torch.class('onmt.CudnnEncoder', 'nn.Container')
+
+--[[ Construct an encoder layer.
+
+Parameters:
+
+  * `inputNetwork` - input module
+  * `rnnClass` - CuDNN recurrent module class
+  * `layers` - number of layers
+  * `inputSize` - input dimension
+  * `rnnSize` - RNN hidden dimension
+  * `dropout` - dropout value
+]]
+function CudnnEncoder:__init(inputNetwork, rnnClass, layers, inputSize, rnnSize, dropout)
+  parent.__init(self)
+
+  self.rnn = rnnClass(inputSize, rnnSize, layers, false, dropout, false)
+  self.inputNetwork = inputNetwork
+
+  self.net = nn.Sequential()
+    :add(self.inputNetwork)
+    :add(nn.Transpose({1, 2})) -- By default, batch is second in CuDNN.
```

Comment:
You are right, it is cleaner. However, I don't think it will be more efficient as cuDNN is doing the transposition internally: https://github.com/soumith/cudnn.torch/blob/master/RNN.lua#L254-L256